test/alternator/test_streams.py

```diff
@@ -0,0 +1,373 @@
+
+            while True:
+                shards = desc['StreamDescription']['Shards']
+
+                for shard in shards:
+                    shard_id = shard['ShardId']
+                    start = shard['SequenceNumberRange']['StartingSequenceNumber']
+                    iter = dynamodbstreams.get_shard_iterator(StreamArn=arn, ShardId=shard_id, ShardIteratorType='AT_SEQUENCE_NUMBER',SequenceNumber=start)['ShardIterator']
+                    iterators.append(iter)
+
+                last_shard = desc["StreamDescription"].get("LastEvaluatedShardId")
+                if not last_shard:
+                    break
+
+                desc = dynamodbstreams.describe_stream(StreamArn=arn, ExclusiveStartShardId=last_shard)
+
+            next_iterators = []
+            while iterators:
+                iter = iterators.pop(0)
+                response = dynamodbstreams.get_records(ShardIterator=iter, Limit=1000)
+                next = response['NextShardIterator']
+
+                if next != '':
+                    next_iterators.append(next)
+
+                records = response.get('Records')
+                print("Query {} -> {}".format(iter, records))
+                if records:
+                    for record in records:
+                        print("Record: {}".format(record))
+                        type = record['eventName']
+                        dynamodb = record['dynamodb']
+                        keys = dynamodb['Keys']
+                        
+                        assert keys.get('p')
+                        assert keys.get('c')
+                        assert keys['p'].get('S')
+                        assert keys['p']['S'] == p
+                        assert keys['c'].get('S')
+                        assert keys['c']['S'] == c
+
+                        if type == 'MODIFY' or type == 'INSERT':
+                            assert dynamodb.get('NewImage')
+                            newimage = dynamodb['NewImage'];
+                            assert newimage.get('a1')
+                            assert newimage.get('a2')
+
+                        if type == 'INSERT' or (type == 'MODIFY' and not has_insert):
+                            assert newimage['a1']['S'] == val
+                            assert newimage['a2']['S'] == val2
+                            has_insert = True
+                            continue
+                        if type == 'MODIFY':
+                            assert has_insert
+                            assert newimage['a1']['S'] == nval
+                            assert newimage['a2']['S'] == nval2
+                            assert dynamodb.get('OldImage')
+                            oldimage = dynamodb['OldImage'];
+                            assert oldimage.get('a1')
+                            assert oldimage.get('a2')
+                            assert oldimage['a1']['S'] == val
+                            assert oldimage['a2']['S'] == val2
```

Comment:
Do secondary indices affect the base table layout in alternator? Are those not separate tables?
I agree that is would be good to force a base table that has both attrs map and real columns (except the keys) and verify this, but I am unsure on how to create this. 
I would maybe want to postpone adding this to the test (added comment).

Reply:
In the current implementation, when secondary indexes are not involved, just the two key columns (partition and sort) are real Scylla columns - all the rest are stuffed into one ":attrs" column (a map). However, with LSI or GSI, we have additional indexed column(s), and this column(s) are real Scylla columns as well (because of limitations of our materialized views implementation, which doesn't know to take values from the map and needs to index a real column). So if the streams code expects that all of the columns except the key are always in ":attrs", it won't work when there are LSI or GSI.
Sure, you can add a TODO and check this later. Even better (but I don't insist) is to write a test, and if it fails on Alternator, leave it as "xfail" - these are much better than a TODO.